johnnyfive/gmail.py

```diff
@@ -0,0 +1,404 @@
+        -------
+        `Any`
+            Uh, the Message object from Gmail... probably just return nothing?
+        """
+        if not add_labels and not remove_labels:
+            print("No labels to change.")
+            return None
+
+        # Convert Label Names to Label IDs
+        add_label_ids, remove_label_ids = [], []
+        if add_labels:
+            for label in add_labels:
+                add_label_ids.append(self._lableId_from_labelName(label))
+        if remove_labels:
+            for label in remove_labels:
+                remove_label_ids.append(self._lableId_from_labelName(label))
+
+        # Build the label dictionary to send to Gmail
+        body = {}
+        if add_label_ids:
+            body['addLabelIds'] = add_label_ids
+        if remove_label_ids:
+            body['removeLabelIds'] = remove_label_ids
+
+        try:
+            # Modify message lables (API: users.messages.modify)
+            return self.service.users().messages().modify(userId='me',
+                                            id=message_id, body=body).execute()
+        # TODO: Actually deal with this exception properly.
+        except Exception as e:
+            print('Gack.')
+            raise e
+
+    def _lableId_from_labelName(self, name):
+        """_lableId_from_labelName Get the Label ID from the Label Name
+
+        _extended_summary_
+
+        Parameters
+        ----------
+        name : `str`
+            Label name
+
+        Returns
+        -------
+        `str`
+            Label ID
+        """
+        # Only do this once
+        if not self.label_list:
+            # Get the list of labels for the "me" account (API: users.labels.list)
+            results = self.service.users().labels().list(userId='me').execute()
+            self.label_list = results.get('labels', [])
+
+        # If there are no labels, return None
+        if not self.label_list:
+            print('Whoops, no labels found.')
+            return None
+
+        label_id = None
+        # Go through the labels, and return the ID matching the name
+        for label in self.label_list:
+            if label['name'] == name:
+                label_id = label['id']
+
+        return label_id
+
+
+# Newer OAUTH Routines =======================================================#
+def setup_gmail():
+    """setup_gmail Initialize the GMail API (via OAuth)
+
+    [extended_summary]
+
+    NOTE: The first time this is run on a machine, it will open a webpage for
+          authorizing the API.  All subsequent runs will be silent.
+
+    Returns
+    -------
+    `googleapiclient.discovery.Resource`
+        The GMail API service object for consumption by other routines
+    """
+    creds = None
+    if os.path.exists(
+            token_fn := utils.Paths.config.joinpath('gmail_token.json') ):
+        creds = Credentials.from_authorized_user_file(token_fn, SCOPES)
+
+    # If there are no (valid) credentials available, lauch browser to log in
```

Comment:
I'm a _little_ worried about this in a container; the ones I run are headless, so this will fail

Reply:
Addressed.